### PR TITLE
Date / Datetime improvements

### LINKED ISSUE #535: vec_size() takes a "long time" for POSIXlt
``` r
library(vctrs)

ct <- new_datetime(0)
lt <- as.POSIXlt(ct)

bench::mark(
  vec_size(ct),
  vec_size(lt)
)
#> # A tibble: 2 x 6
#>   expression        min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>   <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_size(ct)    727ns    944ns  1013283.     6.3KB    101. 
#> 2 vec_size(lt)   18.6µs   21.8µs    42695.    29.5KB     25.6
```

Discovered while repeatedly calling `vec_restore(x, to, R_NilValue)` where `x` was a list holding a POSIXlt and `to` was a data frame. Because no size was passed in, it goes through `df_raw_size(x)`, which in this case calls `vec_size()` on the first element of the list (the POSIXlt). Totally crippled performance 😬 . Luckily I am able to pre-compute the size ahead of time to fix it for my use case.

The main "issue" is that `vec_size()` needs the proxy:
https://github.com/r-lib/vctrs/blob/f736b2353d248820847a09b808cb7370ab8fc2f6/src/size.c#L47

and `vec_proxy.POSIXlt` calls `new_data_frame()`, which is quite slow:

``` r
library(vctrs)

ct <- new_datetime(0)
lt <- as.POSIXlt(ct)

# imitate vec_proxy.POSIXlt
vctrs:::vec_proxy.POSIXlt
#> function(x, ...) {
#>   new_data_frame(unclass(x))
#> }
#> <bytecode: 0x7f99f69716f8>
#> <environment: namespace:vctrs>

# no problem here:
bench::mark(unclass(lt))
#> # A tibble: 1 x 6
#>   expression       min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>  <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 unclass(lt)    308ns    373ns  2350164.        0B     235.

lt_unclassed <- unclass(lt)

# ouch
bench::mark(new_data_frame(lt_unclassed))
#> # A tibble: 1 x 6
#>   expression                       min  median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>                   <bch:t> <bch:t>     <dbl> <bch:byt>    <dbl>
#> 1 new_data_frame(lt_unclassed)  14.9µs  17.6µs    52643.    24.8KB     15.8
```

Perhaps there is no solution, since we have to get the full proxy, but this is definitely something to be aware of when liberally using `vec_size()` and expecting it to have negligible speed impacts.

### LINKED ISSUE #561: vec_c(<POSIXct>, <POSIXct>)
``` r
library(vctrs)

times <- Sys.time() + c(1, 1, 2)
str(vec_ptype(times))
#>  'POSIXct' num(0)
str(vec_ptype(c(times, times)))
#>  'POSIXct' num(0)
str(vec_ptype(vec_c(times, times)))
#>  'POSIXct' num(0) 
#>  - attr(*, "tzone")= chr ""
```

`vec_c()` added an extra `tzone` attribute.

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes #535
Closes #561 

This PR adds a number of date / datetime improvements.

- Constructors are faster
   - And now have a small suite of tests to ensure their behaviors are correct
- Casting in the trivial cases is faster
- R level casting now uses `stop_native_implementation()`
   - Except where we need to fallback for lossy errors
- Proxying is faster
- A few bugs were fixed

Constructor:

```r
bench::mark(
  date = new_date(),
  datetime = new_datetime(),
  check = FALSE
)

# Master
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 date         7.82µs    9.5µs   102145.    8.68KB     81.8
#> 2 datetime        9µs   12.3µs    73829.   15.53KB     29.5

# This PR
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 date          976ns   1.26µs   772226.    6.17KB     77.2
#> 2 datetime     1.14µs   1.33µs   718756.     4.1KB     71.9
```

Proxy:

```r
x <- new_date()
y <- new_datetime()

bench::mark(
  date = vec_proxy(x),
  datetime = vec_proxy(y),
  check = FALSE
)

# Master
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 date         1.71µs   2.03µs   441019.      12KB     44.1
#> 2 datetime    11.64µs  14.83µs    66387.    5.54KB     19.9

# This PR
#> # A tibble: 2 x 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 date         1.71µs   2.29µs   415592.    8.28KB      0  
#> 2 datetime     2.12µs   2.54µs   375744.     8.3KB     37.6
```

Casting (doing nothing)

```r
bench::mark(
  date_date = vec_cast(x, x),
  datetime_datetime = vec_cast(y, y),
  check = FALSE
)

# Master
#> # A tibble: 2 x 6
#>   expression             min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>        <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 date_date           3.05µs   4.06µs   240986.    4.46KB     24.1
#> 2 datetime_datetime  38.29µs  47.62µs    20565.   87.21KB     25.0

# This PR
#> # A tibble: 2 x 6
#>   expression             min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>        <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 date_date            798ns    986ns   940620.        0B     94.1
#> 2 datetime_datetime    862ns   1.01µs   950229.        0B      0
```

Size:

```r
ct <- new_datetime(0)
lt <- as.POSIXlt(ct)

bench::mark(
  vec_size(ct),
  vec_size(lt),
  iterations = 100000
)

# Master
#> # A tibble: 2 x 6
#>   expression        min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>   <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_size(ct)  11.06µs  15.52µs    63956.    11.8KB     69.8
#> 2 vec_size(lt)   2.61µs   3.06µs   308507.    11.1KB     30.9

# This PR
#> # A tibble: 2 x 6
#>   expression        min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>   <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_size(ct)   1.58µs   2.39µs   406129.    14.6KB    106. 
#> 2 vec_size(lt)   2.43µs   2.93µs   314628.    11.1KB     28.3
```

Datetime constructor is stricter now, requires double input (like `new_date()`)

```r
# always required double input
new_date(1L)
#> Error: `x` must be a double vector.

# stricter now
new_datetime(1L)
#> Error: `x` must be a double vector.
```

Bug: Can't convert POSIXct to Date unless it is offset with the correct UTC offset

```r
# Master

x <- as.POSIXct("2019-01-01", tz = "America/New_York")
vec_cast(x, new_date())
#> Error: Can't convert from <datetime<America/New_York>> to <date> due to loss of precision.
#> * Locations: 1

x <- as.POSIXct("2018-12-31 19:00:00", tz = "America/New_York")
vec_cast(x, new_date())
#> [1] "2019-01-01"

# This PR
x <- as.POSIXct("2019-01-01", tz = "America/New_York")
vec_cast(x, new_date())
#> [1] "2019-01-01"

x <- as.POSIXct("2018-12-31 19:00:00", tz = "America/New_York")
vec_cast(x, new_date())
#> Error: Can't convert from <datetime<America/New_York>> to <date> due to loss of precision.
#> * Locations: 1
```

Bug: POSIXlt to POSIXlt conversion didn't change time zones

```r
# Master
x <- as.POSIXlt("2019-01-01", tz = "America/New_York")
to <- as.POSIXlt(new_datetime(tzone = "America/Los_Angeles"))
vec_cast(x, to)
#> [1] "2019-01-01 EST"

# This PR
x <- as.POSIXlt("2019-01-01", tz = "America/New_York")
to <- as.POSIXlt(new_datetime(tzone = "America/Los_Angeles"))
vec_cast(x, to)
#> [1] "2018-12-31 21:00:00 PST"
```

### CODE REVIEW

File: tests/testthat/test-type-date-time.R [hunk 1]
```diff
@@ -153,9 +207,17 @@ test_that("safe casts work as expected", {
...
+  datetime <- as.POSIXct("2018-01-01", tz = "America/New_York")
+
+  expect_identical(vec_cast(vec_cast(date, datetime), date), date)
+  expect_identical(vec_cast(vec_cast(datetime, date), datetime), datetime)
+})
+
 test_that("lossy casts generate error", {
   date <- as.Date("2018-01-01")
-  datetime <- as.POSIXct(date) + c(0, 3600)
+  datetime <- as.POSIXct(as.character(date)) + c(0, 3600)
```
Comment (hadley) [inline tests/testthat/test-type-date-time.R:220]: Why do you need this?

Comment (DavisVaughan) [reply tests/testthat/test-type-date-time.R:220]: It goes along with one of the bug fixes. Compare:

```r
date <- as.Date("2018-01-01")

as.POSIXct(date)
#> [1] "2017-12-31 19:00:00 EST"

as.POSIXct(as.character(date))
#> [1] "2018-01-01 EST"
```

This test expects that when you allow a lossy cast of the `datetime` to `date`, that it gives identical results to `vec_c(date, date)`. If we don't cast to character first to ignore time zone shifts, then with the newly corrected behavior casting to date will give `"2017-12-31"` which won't match `date`

Comment (hadley) [reply tests/testthat/test-type-date-time.R:220]: Oh of course 🙄

**Comment by DavisVaughan** [comment]
```r
library(vctrs)

show_tzone <- function(x) {
  attr(x, "tzone", exact = TRUE)
}

times <- Sys.time()

# Sys.time() returns a POSIXct with no `tzone` attribute
show_tzone(times)
#> NULL

# a `vec_restore()` method now ensures that proxying then restoring to a
# posixct always standardizes the `tzone` attribute. a `""` is always
# added if there was no tzone before.

show_tzone(vec_slice(times, 0))
#> [1] ""

show_tzone(vec_ptype(times))
#> [1] ""

show_tzone(vec_c(times, times))
#> [1] ""
```

Previous behavior on master:

```r
show_tzone(vec_slice(times, 0))
#> NULL

show_tzone(vec_ptype(times))
#> NULL

show_tzone(vec_c(times, times))
#> [1] ""
```

File: R/type-date-time.R [hunk 2]
```diff
@@ -493,12 +498,30 @@ units_union <- function(x, y) {
...
-  }
+date_validate <- function(x) {
+  .Call(vctrs_date_validate, x)
+}
+
+datetime_validate <- function(x) {
+  .Call(vctrs_datetime_validate, x)
+}
+
+# as.character.Date() calls format() which tries to guess a simplified format.
```
Comment (lionel-) [inline R/type-date-time.R:509]: 👍

**Pull request merged**